File: train/preparation.py
"""Fresh development data and frozen-weight comparison; no final-test selection."""
import argparse
import json
import math
from pathlib import Path
import numpy as np
from PIL import Image, ImageDraw, ImageFont, ImageOps
import torch
from train.robustness import read, write, sha
from train.ten import batch
from train.ten_model import load_export
from train.hundred import probabilities

ROOT = Path(__file__).resolve().parents[1]
DATA = ROOT / '.data/preparation'
TEXTS = ['R', 'm', 'vy', 'kt', 'cedar', 'glyphs', 'Silver marks drift', 'Nimble owls turn']


def plans():
    DATA.mkdir(parents=True, exist_ok=True)
    used = {t.lower() for pool in read(ROOT / 'bench/hundred-texts.json').values() for t in pool}
    if any(t.lower() in used for t in TEXTS if len(t)>1): raise ValueError('Development text repeats historical corpus')
    write(DATA / 'plans.json', {'texts':TEXTS,'seed':20260924,'scope':'Development only; alphabet shared, multi-character strings disjoint from historical corpus.'})


def render():
    fonts = [f for f in read(ROOT/'bench/fonts-100.json')['fonts'] if f['split']=='train']
    browser = read(DATA/'browser.json'); raw = (DATA/'browser.gray').read_bytes()
    rows=[]
    with (DATA/'source.gray').open('wb') as out:
        def append(image, row):
            original=image
            for condition in ['clean','small','rotate','neighbor']:
                image=original.copy(); regions=np.array(row['regions'],dtype=float); angle=0
                if condition=='small':
                    image=image.resize((max(1,round(image.width*.45)),max(1,round(image.height*.45))),Image.Resampling.LANCZOS)
                    regions[:,:,0]*=image.width/original.width;regions[:,:,1]*=image.height/original.height
                elif condition=='rotate':
                    angle=6 if len(row['text'])%2 else -6
                    image=image.rotate(-angle,Image.Resampling.BICUBIC,expand=True,fillcolor=255)
                    theta=math.radians(angle);matrix=np.array([[math.cos(theta),-math.sin(theta)],[math.sin(theta),math.cos(theta)]])
                    regions=(regions-[original.width/2,original.height/2])@matrix.T+[image.width/2,image.height/2]
                elif condition=='neighbor':
                    # A visible sliver from a real v in the same font, adjacent to the query.
                    f=ImageFont.truetype(str(ROOT/f'.data/fonts/{row["family"]}.ttf'),row['size'])
                    v=Image.new('L',(row['size']*2,row['size']*2),255);ImageDraw.Draw(v).text((0,0),'v',font=f,fill=0)
                    box=ImageOps.invert(v).getbbox(); ink=ImageOps.invert(original).getbbox()
                    if box and ink:image.paste(v.crop((box[0],box[1],box[0]+2,box[3])),(min(image.width-2,ink[2]+1),max(0,ink[3]-(box[3]-box[1]))))
                rows.append({**row,'condition':condition,'angle':angle,'regions':regions.tolist(),'width':image.width,'height':image.height,'offset':out.tell()})
                out.write(image.tobytes())
        for idx,font in enumerate(fonts):
            path=ROOT/f'.data/fonts/{font["id"]}.ttf'
            if sha(path)!=font['instanceSha256']:raise ValueError('Font changed')
            for i,text in enumerate(TEXTS):
                size=[18,32,56][(idx+i)%3];f=ImageFont.truetype(str(path),size)
                box=f.getbbox(text,anchor='ls');x,y=8-box[0],8-box[1]
                image=Image.new('L',(box[2]-box[0]+16,box[3]-box[1]+16),255)
                ImageDraw.Draw(image).text((x,y),text,font=f,fill=0,anchor='ls')
                regions=[];prefix=''
                for word in text.split(' '):
                    a,b,c,d=f.getbbox(word,anchor='ls');left=x+f.getlength(prefix)
                    regions.append([[left+a,y+b],[left+c,y+b],[left+c,y+d],[left+a,y+d]]);prefix+=word+' '
                append(image,{'family':font['id'],'text':text,'size':size,'renderer':'pillow','regions':regions})
        offset=0
        for row in browser['samples']:
            n=row['width']*row['height']
            if row['offset']!=offset:raise ValueError('Invalid browser offset')
            image=Image.frombytes('L',(row['width'],row['height']),raw[offset:offset+n]);offset+=n
            append(image,{**row,'renderer':'chromium'})
        if offset!=len(raw):raise ValueError('Trailing browser pixels')
    write(DATA/'source.json',{'samples':rows,'sha256':sha(DATA/'source.gray'),'fontsSha256':sha(ROOT/'bench/fonts-100.json'),'generatorSha256':sha(Path(__file__))})
    print(f'{len(rows)} fresh development images',flush=True)


def evaluate():
    torch.set_num_threads(4)
    if not torch.backends.mps.is_available():raise ValueError('MPS unavailable')
    manifest=read(DATA/'prepared.json'); artifact=read(ROOT/'models/hundred/model.json')
    for key,path in [('inputSha256','src/input.mjs'),('lineSha256','src/line.mjs'),('normalizerSha256','src/prepare.mjs')]:
        if manifest[key]!=sha(ROOT/path):raise ValueError('Changed preparation source: '+path)
    model=load_export(artifact).to('mps');families=artifact['fonts']; report={}
    for method,entry in manifest['methods'].items():
        path=DATA/f'{method}.u8'
        if sha(path)!=entry['sha256']:raise ValueError('Changed tensors')
        pixels=np.memmap(path,dtype=np.uint8,mode='r');windows=entry['windows'];logits=[];owners=[]
        with torch.no_grad():
            for start in range(0,len(windows),128):
                chosen=list(range(start,min(start+128,len(windows))))
                logits.append(model(*(t.to('mps') for t in batch(pixels,windows,chosen))).cpu().numpy());owners.extend(windows[i]['source'] for i in chosen)
        if len(set(owners))!=len(manifest['samples']):raise ValueError('Dropped development samples')
        probs=probabilities(np.concatenate(logits),np.array(owners),len(manifest['samples']))
        groups={};predictions=[]
        for sample,p in zip(manifest['samples'],probs):
            label=families.index(sample['family']);n=len(sample['text'].replace(' ',''));length='1' if n==1 else '2-3' if n<=3 else '4-7' if n<=7 else '8+'
            hit=int(p.argmax()==label);top=int(label in p.argsort()[-5:])
            for key in ['all',sample['condition'],'length/'+length,sample['renderer']]:
                g=groups.setdefault(key,{'count':0,'correct':0,'top5':0});g['count']+=1;g['correct']+=hit;g['top5']+=top
            predictions.append({'family':sample['family'],'predicted':families[int(p.argmax())],'correct':bool(hit)})
        for g in groups.values():g['accuracy']=g['correct']/g['count'];g['top5Accuracy']=g['top5']/g['count']
        report[method]={'groups':groups,'preparationMs':entry['timing'],'windows':len(windows)}
        write(DATA/f'{method}-predictions.json',predictions)
        print(method,{k:round(v['accuracy'],4) for k,v in groups.items()},flush=True)
    write(ROOT/'bench/preparation.json',{'modelSha256':sha(ROOT/'models/hundred/model.json'),'manifestSha256':sha(DATA/'prepared.json'),'lineSha256':sha(ROOT/'src/line.mjs'),'scope':'Frozen weights; new synthetic development bank, no final-test claims. Oracle uses renderer word boxes and the known rotation.','methods':report})

if __name__=='__main__':
    p=argparse.ArgumentParser();p.add_argument('command',choices=['plans','render','evaluate']);args=p.parse_args()
    globals()[args.command]()

File: scripts/preparation.mjs
import { readFile, writeFile, mkdir, open } from 'node:fs/promises'
import { createHash } from 'node:crypto'
import { spawnSync } from 'node:child_process'
import { chromium } from 'playwright'
import { prepareInput } from '../src/input.mjs'
import { prepareLine } from '../src/line.mjs'

const verification = process.argv[2] === 'verification'
const dir = verification ? '.data/verification' : '.data/preparation', hash = b => createHash('sha256').update(b).digest('hex')
const read = async path => JSON.parse(await readFile(path))
function python(command) {
  const result = spawnSync(process.execPath, ['scripts/python.mjs', '-m', 'train.preparation', command, ...(verification ? ['--verification'] : [])], { stdio: 'inherit' })
  if (result.error || result.status !== 0) throw result.error || new Error(`Failed ${command}`)
}
await mkdir(dir, { recursive: true })
const command = verification ? 'data' : process.argv[2] || 'all'
if (!['all', 'data', 'prepare', 'evaluate'].includes(command)) throw new Error('Expected data, prepare or evaluate')
if (['all', 'data'].includes(command)) {
  python('plans')
  const texts = (await read(`${dir}/plans.json`)).texts
  const fonts = (await read('bench/fonts-100.json')).fonts.filter(f => f.split === 'train' || (verification && f.split === 'unknown-test'))
  const browser = await chromium.launch(), raw = await open(`${dir}/browser.gray`, 'w'), samples = []
  let offset = 0
  try {
    const page = await browser.newPage()
    for (const [index, font] of fonts.entries()) {
      const bytes = await readFile(`.data/fonts/${font.id}.ttf`)
      if (hash(bytes) !== font.instanceSha256) throw new Error(`Font changed: ${font.id}`)
      const rows = await page.evaluate(async ({ bytes, texts, index }) => {
        const face = await new FontFace('probe', new Uint8Array(bytes).buffer).load()
        document.fonts.clear(); document.fonts.add(face)
        return texts.map((text, i) => {
          const size = [18, 32, 56][(index + i) % 3], pad = 8, canvas = document.createElement('canvas')
          let c = canvas.getContext('2d'); c.font = `${size}px probe`
          const m = c.measureText(text)
          canvas.width = Math.ceil(m.actualBoundingBoxLeft + m.actualBoundingBoxRight) + 2 * pad
          canvas.height = Math.ceil(m.actualBoundingBoxAscent + m.actualBoundingBoxDescent) + 2 * pad
          c = canvas.getContext('2d'); c.font = `${size}px probe`; c.fillStyle = 'white'; c.fillRect(0, 0, canvas.width, canvas.height)
          const x = pad + m.actualBoundingBoxLeft, y = pad + m.actualBoundingBoxAscent
          c.fillStyle = 'black'; c.fillText(text, x, y)
          let prefix = ''
          const regions = text.split(' ').map(word => {
            const n = c.measureText(word), left = x + c.measureText(prefix).width - n.actualBoundingBoxLeft
            prefix += word + ' '
            return [[left, y - n.actualBoundingBoxAscent], [left + n.actualBoundingBoxLeft + n.actualBoundingBoxRight, y - n.actualBoundingBoxAscent], [left + n.actualBoundingBoxLeft + n.actualBoundingBoxRight, y + n.actualBoundingBoxDescent], [left, y + n.actualBoundingBoxDescent]]
          })
          const data = c.getImageData(0, 0, canvas.width, canvas.height).data
          let pixels = ''; for (let i = 0; i < data.length; i += 4) pixels += String.fromCharCode(data[i])
          return { text, size, regions, width: canvas.width, height: canvas.height, pixels: btoa(pixels) }
        })
      }, { bytes: Array.from(bytes), texts, index })
      for (const { pixels, ...row } of rows) {
        const bytes = Buffer.from(pixels, 'base64'); await raw.writeFile(bytes)
        samples.push({ ...row, family: font.id, offset }); offset += bytes.length
      }
    }
    await writeFile(`${dir}/browser.json`, JSON.stringify({ samples, browser: browser.version() }))
  } finally { await raw.close(); await browser.close() }
  python('render')
}
if (['all', 'data', 'prepare'].includes(command)) {
  const source = await read(`${dir}/source.json`), input = await readFile(`${dir}/source.gray`)
  const methods = verification ? ['current', 'deskew-windows'] : ['current', 'groups', 'deskew', 'deskew-windows', 'oracle'], reports = {}
  if (hash(input) !== source.sha256) throw new Error('Changed source pixels')
  for (const method of methods) {
    const output = await open(`${dir}/${method}.u8`, 'w'), windows = [], times = [], angles = []
    let offset = 0
    try {
      for (const [id, sample] of source.samples.entries()) {
        const data = new Uint8Array(sample.width * sample.height * 4), gray = input.subarray(sample.offset, sample.offset + sample.width * sample.height)
        if (gray.length !== data.length / 4) throw new Error('Truncated pixels')
        for (let i = 0; i < gray.length; i++) { data.fill(gray[i], i * 4, i * 4 + 3); data[i * 4 + 3] = 255 }
        const start = performance.now()
        const result = method === 'current' ? prepareInput({ ...sample, data }) : prepareLine({ ...sample, data }, { deskew: method.startsWith('deskew'), sampler: method === 'deskew-windows' ? 'windows' : 'groups', ...(method === 'oracle' ? { angle: sample.angle, regions: sample.regions } : {}) })
        times.push(performance.now() - start); angles.push(result.angle || 0)
        for (const window of result.windows) {
          const bytes = Uint8Array.from(window.pixels, p => Math.round(p * 255))
          await output.writeFile(bytes); windows.push({ source: id, offset, width: window.width, height: window.height }); offset += bytes.length
        }
      }
    } finally { await output.close() }
    const sorted = [...times].sort((a,b) => a-b)
    reports[method] = { windows, angles, sha256: hash(await readFile(`${dir}/${method}.u8`)), timing: { median: sorted[Math.floor(times.length / 2)], p95: sorted[Math.floor(times.length * .95)] } }
    console.log(`${method}: ${windows.length} windows, preparation ${reports[method].timing.median.toFixed(2)} ms median`)
  }
  await writeFile(`${dir}/prepared.json`, JSON.stringify({ ...source, methods: reports, inputSha256: hash(await readFile('src/input.mjs')), lineSha256: hash(await readFile('src/line.mjs')), normalizerSha256: hash(await readFile('src/prepare.mjs')) }))
}
if (['all', 'evaluate'].includes(command)) python('evaluate')
